### PR TITLE
Fix event parsing brainvision

### LINKED ISSUE #5420: BUG: BrainVision reads more events than it says it does.
According to the docstring, the BrainVision reader will only read "Stimulus" events, which are typically in the shape `r"(S\d{1,3})"`, so an S and then 1 to 3 digits. All other events will have to be specified using the `event_id` dictionary:
https://github.com/mne-tools/mne-python/blob/beda451422b4d8016643438ce7e998c76cbd2f59/mne/io/brainvision/brainvision.py#L233-L239

However, the BrainVision reader parses all events that end with a number, due to this regexp:
https://github.com/mne-tools/mne-python/blob/beda451422b4d8016643438ce7e998c76cbd2f59/mne/io/brainvision/brainvision.py#L325

This leads to issues when e.g., a "Comment" event is in the markers such as `Mk9=Comment,Discontinuit? [B] 4156 ms,136544,1,0` as in #5356. See [regexp101](https://regex101.com/r/VAqMnU/1) to play around with it.

That comment will then be automatically added to the events. Crucially, its trigger value will be the number that was found, in this case `4156`. That number originally referred to milliseconds. In this particular case the error is not too bad and will just lead to a nonsensical event value. However imagine that it was `254ms`: as BrainVision uses trigger values from 0 to 255, the 254 falls in the range of sensible event values and thus has the potency to be falsely added as a "true event" when it really is just a misread comment.

### PR DESCRIPTION
**Pull request by sappelhoff**
fixes #5420 

TL;DR: We now by default parse events that are stimuli `S###` or responses `R###`, covering the bulk of events in BrainVision files. All other events have to be specified using `event_id`.

-----------------------
The changes to the tests were much more minimal than I expected. Basically I:
1. added `'0  1': 1` to the `event_id` so that the *optic event* in our test files gets parsed (before this bugfix it was wrongly parsed by default)
2. added the `event_id` argument to a `read_raw_brainvision` call where previously the `event_id` was apparently forgotten.

Next to that, I noticed yet another bug that we did not know of before, look at this code:

https://github.com/mne-tools/mne-python/blob/62ecf9e482bf84a7f58c5fefff766336068f8bce/mne/io/brainvision/brainvision.py#L316-L332

The `trig_shift_by_type` is only applied if an `mdesc` was not found in `event_id`. However of course we want to apply the `trig_shift_by_type` also when we have an automatically parsed event type (such as S### and R### from now on). I fixed that by removing one level of indentation.

### CODE REVIEW

File: mne/io/brainvision/brainvision.py [hunk 1]
```diff
@@ -331,19 +331,24 @@ def _read_vmrk_events(fname, event_id=None, trig_shift_by_type=None):
             if cur_shift is not None:
                 trigger += cur_shift
             else:
+                # The trigger has been deliberately shifted to None. Do not
+                # add this to "dropped" so we do not warn about something
+                # that was done deliberately. Just continue with next item.
                 trigger = None
+                continue
```
Comment (sappelhoff) [inline mne/io/brainvision/brainvision.py]: This `continue` is needed so that when a user specifies to shift a trigger to `None` (i.e., deleting it) there is no warning raised that this trigger will be dropped. With the `continue`, a warning is only raised if a trigger is dropped unintentionally.

See also the contingent changes in the tests that expected warnings that are no longer raised now.

**Comment by sappelhoff** [comment]
After a review, I think this is ready to be merged
@larsoner 
@agramfort

**Comment by jona-sassenhagen** [comment]
Do we really want S and R triggers to be parsed the same?

**Comment by sappelhoff** [comment]
> Do we really want S and R triggers to be parsed the same?

It would be convenient but sometimes people would have to specify an offset for triggers using the `trig_shift_by_type` to avoid that trigger values get "merged", as for example in the case when we have a `R123` and an `S123`.

The alternative is to only ever parse stimulus events and let everything else be specified by `event_id`. I'd be fine with that as well.

**Comment by jona-sassenhagen** [comment]
>The alternative is to only ever parse stimulus events and let everything else be specified by event_id. I'd be fine with that as well.

That's how I originally thought about it, assuming most people probably only have a handful of R triggers, but might have many S triggers.

Otherwise, definitely need a big fat warning if triggers would be merged between S and R.

**Comment by sappelhoff** [comment]
> ... call the issue "add trigger warnings"

actually ... I am kind of tending towards only reading S### events now ...

no warnings necessary in that case

**Comment by jona-sassenhagen** [comment]
Might wanna include an extra a warning if RXXX events are detected.

**Comment by sappelhoff** [comment]
> Might wanna include an extra a warning if RXXX events are detected. 

We have that already :-) 

`Currently, 2 trigger(s) will be dropped, such as [comment using [square] brackets, R255]. Consider using ``event_id`` to parse triggers that do not follow the 'S###' pattern`

**Comment by jona-sassenhagen** [comment]
Yeah I mean, add an EXTRA warning if dropped triggers include RXX triggers.

It's not so bad if 'boundary event' triggers etc are lost, but if RXX triggers are lost, that's probably bad.

**Comment by sappelhoff** [comment]
> Yeah, I know, I wrote that code. I mean, add an EXTRA warning if dropped triggers include RXX triggers. It's not so bad if 'boundary event' triggers etc are lost, but if RXX triggers are lost, that's probably bad.

I see ... extra diligent ;-) I'd rather finish this issue up though because I have a feeling my PRs tend to grow and grow (probably everybody's PRs). After this one is merged, we can then make a new *easy* issue.

**Comment by jona-sassenhagen** [comment]
Okay makes sense.

File: mne/io/brainvision/brainvision.py [hunk 2]
```diff
@@ -321,27 +323,32 @@ def _read_vmrk_events(fname, event_id=None, trig_shift_by_type=None):
...
+                trigger += cur_shift
+            else:
+                # The trigger has been deliberately shifted to None. Do not
+                # add this to "dropped" so we do not warn about something
+                # that was done deliberately. Just continue with next item.
+                trigger = None
+                continue
         # FIXME: ideally, we would not use the middle column of the events
         # array to store the duration. A better solution would be using
         # annotations.
```
Comment (agramfort) [inline mne/io/brainvision/brainvision.py]: this comment is still highly relevant...

next PR on this will not change the way we do things?

Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: The FIXME comment seems unrelated to the current PR. A fix for that would not change the way we read events.

The comment about "we do not warn if an event is deliberately dropped" describes my change (inserting the `continue`), which I think is a good idea.

**Comment by agramfort** [comment]
just to illustrate what I mean by using Annotations for brainvision reader see:

https://github.com/agramfort/mne-python/commit/f5f203edbd5de0834ddbcf260a98df1be4133564

@sappelhoff does it make sense to you?

cc @massich

**Comment by sappelhoff** [comment]
What would be the advantage of this refactoring?

**Comment by agramfort** [comment]
we have different problems:

- BrainVision reader hacking the 2nd column of events to store the duration of events (only Annotations have a notion of duration)
- Adding stim channels slows down the IO as it means some reallocation to add a row.
- we don't have the same semantic of convertion from description of "events/annotations" in the different readers. EEGLAB uses event_id_func and event_id while brainvision uses trig_shift_by_type and EDF uses annot and annotmap.

honestly there is a common denominator here  which is the notion of Annotations with onset, duration and string description from which we want to make events and event_id to match the semantic of MNE objects. For me the readers should not be that smart and attach an annotations obj as attribute to the raw files and then the clever function should allow to go from this annotations instance to events.

does it make sense?

**Comment by jona-sassenhagen** [comment]
Both EEGLAB and BV use event_id actually.

I agree standardisation is good.

**Comment by jona-sassenhagen** [comment]
EEGLAB uses `event_id_func` and BV uses `trig_shift_by_type` for essentially the same purpose.

**Comment by jona-sassenhagen** [comment]
Maybe it would be preferable to avoid *necessarily* burdening BV (and EEGLAB) people with understanding the Annotations concept. Many ERP people use these formats, and can be ... not particularly tech savvy.

So if it's all transparent in the background, ok. But if they need to understand it, I would prefer avoiding that.

**Comment by agramfort** [comment]
yes it should be transparent.

My idea is to add the annotations attribute and then when you do now
mne.find_events(raw)
you would use mne.find_events also but it would use the annotations and
have a parameter a similar to event_id_func or trig_shift_by_type

makes sense?

**Comment by jona-sassenhagen** [comment]
Yes, makes sense to me. It might even be preferable, because the stim channel concept is IME confusing for many EEG people.

**Comment by sappelhoff** [comment]
> Yes, makes sense to me. It might even be preferable, because the stim channel concept is IME confusing for many EEG people.

+1

File: mne/io/brainvision/brainvision.py [hunk 3]
```diff
@@ -321,27 +323,32 @@ def _read_vmrk_events(fname, event_id=None, trig_shift_by_type=None):
             trigger = event_id[mdesc]
         else:
             try:
-                trigger = int(re.findall(r'[A-Za-z]*\s*?(\d+)', mdesc)[0])
+                trigger = int(re.findall(r'S([\s\d]{2}\d{1})', mdesc)[0])
```
Comment (jona-sassenhagen) [inline mne/io/brainvision/brainvision.py]: yyyyyeah. Is this intentional?

Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: yes, that's intentional - we are matching the BrainVision triggers, which are always in the format specified there.

"S" and then an up to 3 digit integer where the empty spaces are potentially padded by spaces. E.g., 1 would become `S  1` 

Am I missing an obvious flaw?

Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: Well, I do have a bunch of vmrks with no S ... they're from a lab using PyCorder iirc.

Comment (sappelhoff) [reply mne/io/brainvision/brainvision.py]: So there was no conversion from some format into BrainVision that could be blamed? Would be great to know for sure :-) 

In any case, that problem is easily fixed by using the `event_id`. 

What this PR did was to adjust MNEs behavior to what it always said it would do (which is "read events that follow the S### pattern"). Are you voting on changing that?

One further information: With the BIDS-EEG team we have recently reached out to Brain Products and asked them to release a documentation of the BrainVision format to the community and they agreed, work is on the way. I suggest that we wait for that spec and then check how we can improve our BV reader.

Comment (massich) [reply mne/io/brainvision/brainvision.py]: where did you see it broken?

Comment (jona-sassenhagen) [reply mne/io/brainvision/brainvision.py]: >where did you see it broken?

What do you mean?

**Comment by sappelhoff** [comment]
> I checked with the lab and it seems the files came out of Pycorder in that format - no S, just spaces and integers. I think we should support that.

I agree - but we should take extra care to really limit it to `S###` or `###` then, to avoid the bugs we had (at least risked, if not encountered) before

**Comment by jona-sassenhagen** [comment]
Makes sense. Could I actually bother you with doing it? As I said, regex give me headaches ...

**Comment by sappelhoff** [comment]
> Makes sense. Could I actually bother you with doing it? As I said, regex give me headaches ...

- Perhaps you could make an issue?
- Should we maybe wait for the documentation by Brain Products? (Could be out anytime between now and estimated 2 months or so?)
- I think it would be good if you could figure out first, why the `event_id` parsing is not working in that case --> perhaps we have to fix a bug in addition

**Comment by jona-sassenhagen** [comment]
Yes, no, yes :) It's not unlikely many more people using Pycorder will be impacted by this issue, and I'd prefer to reduce this window.

File: mne/io/brainvision/brainvision.py (hunk 3 above)
Comment (massich) [reply mne/io/brainvision/brainvision.py]: > Well, I do have a bunch of vmrks with no S ... they're from a lab using PyCorder iirc.

I mean do we have a test for that? or is it that we are missing the data?
